src/translate_scale.rs

```diff
@@ -0,0 +1,158 @@
+//! A transformation that includes both scale and translation.
+
+use std::ops::{Add, AddAssign, Mul, MulAssign, Sub, SubAssign};
+
+use crate::{Affine, Circle, Point, Rect, Vec2};
+
+/// A transformation including scaling and translation.
+#[derive(Clone, Copy, Debug)]
+pub struct TranslateScale {
+    translation: Vec2,
+    scale: f64,
+}
+
+impl TranslateScale {
+    /// Create a new transformation from translation and scale.
+    #[inline]
+    pub fn new(translation: Vec2, scale: f64) -> TranslateScale {
+        TranslateScale { translation, scale }
+    }
+
+    /// Create a new transformation with scale only.
+    #[inline]
+    pub fn scale(s: f64) -> TranslateScale {
+        TranslateScale::new(Vec2::ZERO, s)
+    }
+
+    /// Create a new transformation with translation only.
+    #[inline]
+    pub fn translate(t: Vec2) -> TranslateScale {
+        TranslateScale::new(t, 1.0)
+    }
+
+    /// Decompose transformation into translation and scale.
+    pub fn as_tuple(self) -> (Vec2, f64) {
+        (self.translation, self.scale)
+    }
+}
+
+impl Default for TranslateScale {
+    #[inline]
+    fn default() -> TranslateScale {
+        TranslateScale::scale(1.0)
+    }
+}
+
+impl From<TranslateScale> for Affine {
+    fn from(ts: TranslateScale) -> Affine {
+        let TranslateScale { translation, scale } = ts;
+        Affine::new([scale, 0.0, 0.0, scale, translation.x, translation.y])
+    }
+}
+
+impl Mul<Point> for TranslateScale {
+    type Output = Point;
+
+    #[inline]
+    fn mul(self, other: Point) -> Point {
+        (self.scale * other.to_vec2()).to_point() + self.translation
+    }
+}
+
+impl Mul for TranslateScale {
+    type Output = TranslateScale;
+
+    #[inline]
+    fn mul(self, other: TranslateScale) -> TranslateScale {
+        TranslateScale {
+            translation: self.translation + self.scale * other.translation,
+            scale: self.scale * other.scale,
+        }
+    }
+}
+
+impl MulAssign for TranslateScale {
+    #[inline]
+    fn mul_assign(&mut self, other: TranslateScale) {
+        *self = self.mul(other);
+    }
+}
+
+impl Mul<TranslateScale> for f64 {
```

Comment:
Same issue of commutativity as above. I'd like `ts * s` to be the same as `ts * TranslateScale::scalar(s)`. The latter multiplies the scale but does *not* affect the translation, whereas `s * ts` does both.

Is this confusing? To me, following the way math does it is clear, but that might not be true for all users. Perhaps it needs more docstring to emphasize these subtle distinctions.

Reply:
scalars feel like a distinct case, and one where I don't have any strongly held opinion. My understanding of how it would work is that it would just ignore the vector component, as you describe. I'm just unsure if this is useful or helpful behaviour.